### PR TITLE
new way of passing dataloaders

### PR DESCRIPTION
**Pull request by SkafteNicki**
# Before submitting

- [x] Was this discussed/approved via a Github issue? (no need for typos, doc improvements)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md)?
- [x] Did you make sure to update the docs?   
- [ ] Did you write any new necessary tests?  

## What does this PR do?
This is a initial solution to issue https://github.com/PyTorchLightning/pytorch-lightning/issues/662. This allows people to pass a `train_dataloader`, `val_dataloader` and `test_dataloader` to the `.fit()` method of `Trainer`. As mentioned in the issue, this is a more familiar interface for people coming from `keras` and `scikit-learn`. 

What needs to be discussed is probably in the case where a dataloader is defined both in the model and passed to `fit`. Right now the `fit` method will only use the dataloader defined in the model, and prompt the user that it is skipping the dataloader passed to `fit`. For `train_dataloader` there needs to be some kind of tiebreaker, however since both `val_dataloader` and `test_dataloader` can be a list of multiple dataloader a option here would be to concatenate dataloaders defined in the model and those passed to `fit`. 

## PR review    
Anyone in the community is free to review the PR once the tests have passed.     
If we didn't discuss your PR in Github issues there's a high chance it will not be merged.

## Did you have fun?
yes!

### CODE REVIEW

File: pytorch_lightning/trainer/trainer.py [hunk 1]
```diff
@@ -713,17 +713,64 @@ def tng_tqdm_dic(self):
...
+                                 'of torch.utils.data.DataLoader')
+
+        if val_dataloader:
+            if isinstance(val_dataloader, torch.utils.data.DataLoader):
+                if model.val_dataloader() is None:
+                    model.val_dataloader = lambda: val_dataloader
+                else:
+                    logging.info('Model has predefined val_dataloader, '
+                                 'will skip the val_dataloader passed to fit method ')
+            else:
```
Comment (Borda) [inline pytorch_lightning/trainer/trainer.py]: you may merge this with the first if, so `elif val_dataloader:`

File: pytorch_lightning/trainer/trainer.py [hunk 2]
```diff
@@ -713,17 +713,64 @@ def tng_tqdm_dic(self):
...
+                DataLoader with validation samples. If the model has
+                a predefined val_dataloader method this will be skipped
+
+            test_dataloader (:class:`.torch.utils.data.DataLoader`): A Pytorch
+                DataLoader with test samples. If the model has a
+                a predefined test_dataloader method this will be skipped
 
-            trainer.fit()
         """
+        if train_dataloader:
```
Comment (Borda) [inline pytorch_lightning/trainer/trainer.py]: I would make it a function since the three are almost the same:
```
def _set_dataloader(dataloader, attribute):
    if isinstance(dataloader, torch.utils.data.DataLoader):
        if getattr(model, attribute) is None:
            setattr(model, attribute, lambda: dataloader)
        else:
            logging.info(f'Model has predefined `{attribute}`, '
                         'will skip the `{attribute}` passed to fit method.')
    elif dataloader:
        raise ValueError(f'`{attribute}` needs to be an instance of `torch.utils.data.DataLoader`')
    
_set_dataloader(train_dataloader, 'train_dataloader')
...
```

**Comment by SkafteNicki** [comment]
@Borda, took your changes into consideration, and made it all into a function. In addition I added:
1) both `val_dataloader` and `test_dataloader` can now be lists of pytorch dataloaders, similar to how it works when they are defined as part of the model
2) i changed the way that the dataloaders are checked whether they are defined by the user or directly inherited from the base class, such that it does not actually involve evaluating the dataloaders
3) to do 2. I had to fix a small bug in `pl.data_loader` decorator, where the `@wraps(fn)` decorator did not do what it was meant to do

**Comment by williamFalcon** [comment]
@SkafteNicki this has been brought up before. Can you tell me again why we should enable this? I do agree that it might be easier to allow this model to server arbitrary data. BUT this will come at the expense of breaking the LightningTemplate and having data + transforms defined in unconventional ways that don't lend themselves to reproducibility

**Comment by SkafteNicki** [comment]
As I initially wrote I think the current dataloader setup is not intuitive for newcomers. It is easy to learn, but it is just different than other common frameworks. 

My personal use case is where i have two models, that needs to be trained on different training sets but evaluated on the same val/test set. For me it makes sense to then let the training dataloaders be defined as part of the model, but pass the val/test dataloaders to the trainer.

Lastly, I do not think that there is anything wrong with having this as additional option for parsing data. It should in no way replace the current way of doing it. 

If this means that we are giving up too much reproducibility, then maybe it is not worth it.

**Comment by williamFalcon** [comment]
@SkafteNicki makes sense! let's do it.
mind fixing the tests?

File: pytorch_lightning/trainer/trainer.py [hunk 3]
```diff
@@ -713,17 +713,38 @@ def tng_tqdm_dic(self):
...
         Runs the full optimization routine.
 
-        Example::
+        Args:
+            model (LightningModule): Model to fit.
+            Example::
 
-            trainer = Trainer()
-            model = LightningModule()
+                trainer = Trainer()
```
Comment (williamFalcon) [inline pytorch_lightning/trainer/trainer.py]: write two examples:

Example 1: no dataloaders passed, then say that it will use the train_dataloader, etc... calls.

Example 2: with the .fit(...)

In addition, elaborate as to why we might want to do one  vs the other

**Comment by williamFalcon** [comment]
@SkafteNicki we also need a test for this. Add it under the tests for  trainer.

Awesome addition!

**Comment by williamFalcon** [comment]
@SkafteNicki mind finishing this over the next day or so? that way we can add to next release. awesome feature!

**Comment by SkafteNicki** [comment]
@williamFalcon i will try to look at this tomorrow, will hopefully get it all to work and implement the appropriate tests. Will let you know as soon as I am done.
When is next release date?

**Comment by williamFalcon** [comment]
ideally next week (we were on a cycle of every 6th), but we got behind in dec

**Comment by Borda** [comment]
https://github.com/PyTorchLightning/pytorch-lightning/milestone/4

**Comment by williamFalcon** [comment]
@SkafteNicki love this feature. can we please get this merged?

**Comment by williamFalcon** [comment]
@SkafteNicki can you add a test and documentation for this please?

**Comment by SkafteNicki** [comment]
@williamFalcon test and documentation should be added now. The failing CI testing seems to be a windows specific problem not related to this PR. Two notes on changes since last time:
* I removed the `@abstractmethod` decorator from `train_dataloader(self)` since this does not need to be define by the user anymore
* I did a bit of renaming of the `LightningTestModelBase` to get one base test model with and without a predefined `train_dataloader(self)`. Please review this part

File: tests/models/base.py [hunk 4]
```diff
@@ -36,7 +36,7 @@ def __init__(self, root, train=True, transform=None, target_transform=None,
         self.targets = self.targets[:num_samples]
 
 
-class LightningTestModelBase(LightningModule):
+class _LightningTestModelBase(LightningModule):
```
Comment (Borda) [inline tests/models/base.py]: pls do not use _ for class name

File: tests/test_trainer.py [hunk 5]
```diff
@@ -367,7 +368,7 @@ class CurrentTestModel(
 
     # verify there are 2 val loaders
     assert len(trainer.get_val_dataloaders()) == 2, \
-        'Multiple val_dataloaders not initiated properly'
+        'Multiple val_dataloade not initiated properly'
```
Comment (Borda) [inline tests/test_trainer.py]: typo

File: tests/test_trainer.py [hunk 6]
```diff
@@ -411,5 +412,103 @@ class CurrentTestModel(
     trainer.test()
 
 
+def test_dataloaders_passed_to_fit(tmpdir):
+    """Verify that dataloaders can be passed to fit"""
+    tutils.reset_seed()
+
+    class CurrentTestModel(
```
Comment (Borda) [inline tests/test_trainer.py]: `LocalTestModel` ?

File: tests/test_trainer.py [hunk 7]
```diff
@@ -411,5 +412,103 @@ class CurrentTestModel(
...
+
+    # logger file to get meta
+    trainer_options = dict(
+        default_save_path=tmpdir,
+        max_epochs=1,
+        val_percent_check=0.1,
+        train_percent_check=0.2
+    )
+
+    # fit model
```
Comment (Borda) [inline tests/test_trainer.py]: pls split these test as separated tests so it is easier to debug it in future is one of these cases fails...

File: tests/test_trainer.py (hunk 6 above)
Comment (SkafteNicki) [reply tests/test_trainer.py]: Since all the other test call it `CurrentTestModel`, I think it should remain as such

File: tests/models/base.py (hunk 4 above)
Comment (SkafteNicki) [reply tests/models/base.py]: Renamed to `TestModelBase`.

File: pytorch_lightning/trainer/trainer.py [hunk 8]
```diff
@@ -935,3 +971,34 @@ def test(self, model=None):
...
+    # Check if attribute comes directly from base class or
+    # derived in user subclass
+    if LightningModule.__qualname__ in getattr(model, attribute).__qualname__:
+        # Val and test should be list of dataloaders
+        dataloader = dataloader if attribute == 'train_dataloader' or \
+            (attribute != 'train_dataloader' and isinstance(dataloader, list)) else [dataloader]
+
+        # Check if input is correct
+        if isinstance(dataloader, torch.utils.data.DataLoader) or \
+           (isinstance(dataloader, list) and all(isinstance(d, torch.utils.data.DataLoader) for d in dataloader)):
```
[the remark this answers is not in the document]
Comment (SkafteNicki) [reply pytorch_lightning/trainer/trainer.py]: Yes, both the `val_dataloader` and `test_dataloader` arguments can be lists of pytorch dataloaders, similar to how they can also be lists when defined inside the model

File: pytorch_lightning/trainer/trainer.py [hunk 9]
```diff
@@ -935,3 +971,34 @@ def test(self, model=None):
...
+    r''' Check dataloaders passed to .fit() method if they are pytorch DataLoader
+         objects and whether or not we should overright the corresponding dataloader
+         in the model '''
+
+    # Check if attribute comes directly from base class or
+    # derived in user subclass
+    if LightningModule.__qualname__ in getattr(model, attribute).__qualname__:
+        # Val and test should be list of dataloaders
+        dataloader = dataloader if attribute == 'train_dataloader' or \
+            (attribute != 'train_dataloader' and isinstance(dataloader, list)) else [dataloader]
```
[the remark this answers is not in the document]
Comment (SkafteNicki) [reply pytorch_lightning/trainer/trainer.py]: Only `val_dataloader` and `test_dataloader` should be converted into lists (if not already one). `train_dataloader` can only be single dataloader.

File: pytorch_lightning/trainer/trainer.py [hunk 10]
```diff
@@ -935,3 +971,34 @@ def test(self, model=None):
...
+
+        # Check if input is correct
+        if isinstance(dataloader, torch.utils.data.DataLoader) or \
+           (isinstance(dataloader, list) and all(isinstance(d, torch.utils.data.DataLoader) for d in dataloader)):
+
+            # Overwrite abstract methods
+            dl = lambda: dataloader
+            dl.__name__ = attribute
```
[the remark this answers is not in the document]
Comment (SkafteNicki) [reply pytorch_lightning/trainer/trainer.py]: basically its mostly for convenience, such that the attribute is not called `lamda`-something. The lambda is necessary for the attribute to behave like a class method.

File: pytorch_lightning/trainer/trainer.py [hunk 11]
```diff
@@ -734,17 +734,53 @@ def tng_tqdm_dic(self):
...
+
+            # Option 2
+            # Dataloaders passed to fit method
+            train, val, test = DataLoader(...), DataLoader(...), DataLoader(...)
+            trainer = Trainer()
+            model = LightningModule()
+            trainer.fit(model, train_dataloader=train,
+                        val_dataloader=val, test_dataloader=test)
+
+            # Option 1 & 2 can be mixed, for example the training set can be
```
[the remark this answers is not in the document]
Comment (SkafteNicki) [reply pytorch_lightning/trainer/trainer.py]: Basically just followed the style of the `.test()` class method

File: tests/test_trainer.py [hunk 12]
```diff
@@ -411,5 +412,162 @@ class CurrentTestModel(
...
+    fit_options = dict(val_dataloader=model._dataloader(train=False))
+    results = trainer.fit(model, **fit_options)
+
+    # fit model
+    trainer = Trainer(**trainer_options)
+    fit_options = dict(val_dataloader=model._dataloader(train=False),
+                       test_dataloader=model._dataloader(train=False))
+    results = trainer.fit(model, **fit_options)
+    assert len(trainer.get_val_dataloaders()) == 1, \
+        'val_dataloaders not initiated properly'
```
Comment (Borda) [inline tests/test_trainer.py]: for the variable names use `
```
"`val_dataloaders` not initiated properly"
```
also, I would consider printing in the message the loader content...

File: pytorch_lightning/trainer/trainer.py [hunk 13]
```diff
@@ -935,3 +971,34 @@ def test(self, model=None):
...
+           (isinstance(dataloader, list) and all(isinstance(d, torch.utils.data.DataLoader) for d in dataloader)):
+
+            # Overwrite abstract methods
+            dl = lambda: dataloader
+            dl.__name__ = attribute
+            setattr(model, attribute, dl)
+
+        elif not dataloader and dataloader != [None]:
+            raise ValueError(f'`{attribute}` needs to be an instance of'
+                             '`torch.utils.data.DataLoader` or a list of, instead got'
```
[the remark this answers is not in the document]
Comment (Borda) [reply pytorch_lightning/trainer/trainer.py]: @SkafteNicki ^^
```
"`torch.utils.data.DataLoader` or a list of dataloaders, instead got %r" % dataloader
```

File: tests/models/__init__.py [hunk 14]
```diff
@@ -2,7 +2,8 @@
 
 import torch
 
-from .base import LightningTestModelBase
+from .base import (LightningTestModelBase,
```
Comment (Borda) [inline tests/models/__init__.py]: one line?

**Comment by Borda** [comment]
@SkafteNicki add change to CHANGELOG.md

**Comment by SkafteNicki** [comment]
@Borda I have updated code and CHANGELOG per your request

**Comment by Borda** [comment]
@SkafteNicki great job, Thx!
